File: 07_local_volatility_and_dupire/cpp/SwaptionPriceSurface.cpp
#include "SwaptionPriceSurface.hpp"

/*
    Constructor
*/

SwaptionPriceSurface::SwaptionPriceSurface(
    YieldCurve* curve_,
    VolSurface* volSurface_,
    double tenor_,
    double notional_,
    bool payer_
){

    curve = curve_;
    volSurface = volSurface_;

    tenor = tenor_;
    notional = notional_;
    payer = payer_;

    expiries = volSurface->expiries;
    strikes = volSurface->strikes;
}

/*
    Build Black swaption price matrix.

    For each expiry:
    - construct forward-starting swap
    - read Black vol from implied vol surface
    - price payer/receiver swaption using Chapter 06 BlackSwaption
*/

std::vector<std::vector<double>>
SwaptionPriceSurface::build(){

    prices = std::vector<std::vector<double>>(
        expiries.size(),
        std::vector<double>(strikes.size(), 0.0)
    );

    for(size_t i = 0; i < expiries.size(); ++i){

        double expiry = expiries[i];

        ForwardStartingSwap underlyingSwap(
            notional,
            //0.0,                 // fixed rate not needed for forward swap rate
            expiry,
            expiry + tenor,
            2
        );

        for(size_t j = 0; j < strikes.size(); ++j){

            double strike = strikes[j];

            double vol = volSurface->getVol(
                expiry,
                strike
            );

            BlackSwaption swaption(
                notional,
                strike,
                expiry,
                vol,
                underlyingSwap,
                payer
            );

            prices[i][j] = swaption.price(
                *curve
            );
        }
    }

    return prices;
}

File: 07_local_volatility_and_dupire/python/swaption_price_surface.py
"""
swaption_price_surface.py

Build Black swaption price surface C(T,K).

This module connects:
- YieldCurve
- ForwardStartingSwap
- BlackSwaption
- VolSurface
"""

import numpy as np

from forward_swap import ForwardStartingSwap
from black_swaption import BlackSwaption


class SwaptionPriceSurface:
    """
    Builds swaption price surface C(T,K).
    """

    def __init__(
        self,
        curve,
        vol_surface,
        tenor=5.0,
        notional=1.0,
        payer=True
    ):

        self.curve = curve
        self.vol_surface = vol_surface
        self.tenor = tenor
        self.notional = notional
        self.payer = payer

        self.expiries = vol_surface.expiries
        self.strikes = vol_surface.strikes

        self.price_matrix = None

    def build(self):
        """
        Build rectangular Black swaption price matrix.
        """

        prices = np.full(
            (len(self.expiries), len(self.strikes)),
            np.nan
        )

        for i, expiry in enumerate(self.expiries):

            underlying_swap = ForwardStartingSwap(
                notional=self.notional,
                fixed_rate=None,
                start=expiry,
                end=expiry + self.tenor,
                payment_frequency=2
            )

            for j, strike in enumerate(self.strikes):

                vol = self.vol_surface.get_vol(
                    expiry,
                    strike
                )

                swaption = BlackSwaption(
                    notional=self.notional,
                    strike=strike,
                    expiry=expiry,
                    underlying_swap=underlying_swap,
                    volatility=vol,
                    payer=self.payer
                )

                prices[i, j] = swaption.price(
                    self.curve
                )

        self.price_matrix = prices

        return prices